rclpy/test/test_time.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from rclpy.clock import ClockType
+from rclpy.duration import Duration
+from rclpy.time import Time
+
+from test_msgs.msg import Builtins
+
+
+class TestTime(unittest.TestCase):
+
+    def test_time_construction(self):
+        time = Time()
+        assert time.nanoseconds == 0
+
+        time = Time(seconds=1, nanoseconds=5e8, clock_type=ClockType.SYSTEM_TIME)
+        assert time.nanoseconds == 1500000000
+        assert time.clock_type == ClockType.SYSTEM_TIME
+
+        with self.assertRaises(OverflowError):
+            time = Time(nanoseconds=2**64)
+        time = Time(nanoseconds=2**64 - 1)
+        assert time.nanoseconds == 2**64 - 1
+
+        with self.assertRaises(ValueError):
+            time = Time(seconds=-1)
+        with self.assertRaises(ValueError):
+            time = Time(nanoseconds=-1)
+        with self.assertRaises(TypeError):
+            time = Time(clock_type='SYSTEM_TIME')
+
+    def test_duration_construction(self):
+        duration = Duration()
+        assert duration.nanoseconds == 0
+
+        duration = Duration(seconds=1, nanoseconds=5e8)
+        assert duration.nanoseconds == 1500000000
+
+        with self.assertRaises(OverflowError):
+            duration = Duration(nanoseconds=2**64)
+        duration = Duration(nanoseconds=2**64 - 1)
+        assert duration.nanoseconds == 2**64 - 1
+
+        with self.assertRaises(ValueError):
+            duration = Duration(seconds=-1)
+        with self.assertRaises(ValueError):
+            duration = Duration(nanoseconds=-1)
+
+    def test_time_operators(self):
+        # Check that modifying the original doesn't affect a copy
```

Comment:
This code doesn't make a copy; though, I don't think a copy test is necessary.

```python
        # make a reference to a new instance of Time() (instance 1) called time1
        time1 = Time(nanoseconds=2, clock_type=ClockType.SYSTEM_TIME)
        # make another reference to (instance 1) called time2
        time2 = time1
        # make time 1 a reference to a new instance of Time() (instance 2)
        time1 = Time(nanoseconds=1, clock_type=ClockType.STEADY_TIME)
```

Which is the same as

```python
        time2 = Time(nanoseconds=2, clock_type=ClockType.SYSTEM_TIME)
        time1 = Time(nanoseconds=1, clock_type=ClockType.STEADY_TIME)
```

A copy would need to be created with [copy.copy() or copy.deepcopy()](https://docs.python.org/3.7/library/copy.html). A shallow copy test isn't necessary because `Time` appears to be immutable.

```python
import copy
# make a reference to a new instance of Time() (instance 1) called time1
time1 = Time(nanoseconds=2, clock_type=ClockType.SYSTEM_TIME)
# make a reference to a new instance of Time() (instance 2) called time2 whose members have been shallow copied from time1.
time2 = copy.copy(Time(nanoseconds=2, clock_type=ClockType.SYSTEM_TIME))
# Not implemented, time properties don't have setter's
# time2.nanoseconds = 5
# time2.clock_type = ClockType.ROS_TIME
# If time was mutable this would fail because time2._time_handle and time1._time_handle are references to the same pycapsule
# assert time2.nanosseconds != time1.nanoseconds
```

and a deepcopy test isn't needed because `Time()` doesn't support it (would need to implement `__getstate__()` or `__setstate__()`)
```python
# 
time2 = copy.deepcopy(time1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.6/copy.py", line 169, in deepcopy
    rv = reductor(4)
TypeError: can't pickle PyCapsule objects

Reply:
I'm with you that this wasn't really testing anything of interest; removed in 81d55681d0702450eb0fc005becc37a4a3365a4c thanks!